### PR TITLE
Enable `NDEBUG` in production builds

### PR DESCRIPTION
**Pull request by tomekzaw**
## Summary

Related to #4901.

This PR properly sets `NDEBUG` flag when building in release mode.

References:
* https://github.com/facebook/react-native/pull/36172
* https://github.com/facebook/react-native/pull/38314
* https://github.com/facebook/react-native/pull/36194

### CODE REVIEW

File: android/CMakeLists.txt [hunk 1]
```diff
@@ -26,8 +26,10 @@ if(${IS_REANIMATED_EXAMPLE_APP})
     string(APPEND CMAKE_CXX_FLAGS " -DIS_REANIMATED_EXAMPLE_APP")
 endif()
 
-if(${CMAKE_BUILD_TYPE} STREQUAL "Debug")
+if(${CMAKE_BUILD_TYPE} MATCHES "Debug")
```
Comment (tjzel) [inline android/CMakeLists.txt:29]: What's the motivation behind this change?

**Review by tjzel — APPROVED** [review]

File: android/CMakeLists.txt (hunk 1 above)
Comment (tomekzaw) [reply android/CMakeLists.txt:29]: > Specifies the build type on single-configuration generators (e.g. [Makefile Generators](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#makefile-generators) or [Ninja](https://cmake.org/cmake/help/latest/generator/Ninja.html#generator:Ninja)). Typical values include Debug, Release, RelWithDebInfo and MinSizeRel, but custom build types can also be defined.

https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html

`CMAKE_BUILD_TYPE` also supports custom values so I thought let's support everything that contains "Debug".

Comment (tjzel) [reply android/CMakeLists.txt:29]: :sus:

**Review by piaskowyk — APPROVED** [review]

File: RNReanimated.podspec [hunk 2]
```diff
@@ -8,14 +8,17 @@ assert_no_reanimated2_with_new_architecture(reanimated_package_json)
...
+
 folly_prefix = config[:react_native_minor_version] >= 64 ? 'RCT-' : ''
 folly_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32 -DREACT_NATIVE_MINOR_VERSION=' + config[:react_native_minor_version].to_s
 folly_compiler_flags = folly_flags + ' ' + '-Wno-comma -Wno-shorten-64-to-32'
 boost_compiler_flags = '-Wno-documentation'
 fabric_flags = fabric_enabled ? '-DRCT_NEW_ARCH_ENABLED' : ''
 example_flag = config[:is_reanimated_example_app] ? '-DIS_REANIMATED_EXAMPLE_APP' : ''
 version_flag = '-DREANIMATED_VERSION=' + reanimated_package_json["version"]
-using_hermes = ENV['USE_HERMES'] == nil || ENV['USE_HERMES'] == '1'
+debug_flag = is_release ? '-DNDEBUG' : '-DDEBUG'
```
Comment (tomekzaw) [inline RNReanimated.podspec:21]: Here's how this line looks in react-native:
```rb
ndebug_flag = (is_release ? " -DNDEBUG" : "")
```
